fedimint-client/src/dynamic_config.rs

```diff
@@ -0,0 +1,271 @@
+        }
+    }
+}
+
+async fn update_config_value(
+    db: &Database,
+    db_key_prefix: u8,
+    api: &impl AnyDynApi,
+    fetch_method: &'static str,
+) -> anyhow::Result<()> {
+    let fetched_value = api
+        .request_current_consensus::<serde_json::Value>(
+            fetch_method.to_owned(),
+            ApiRequestErased::default(),
+        )
+        .await?;
+
+    let db_key = DynConfigKey {
+        db_key_prefix,
+        fetch_method: fetch_method.to_owned(),
+    };
+    let db_value = DynConfigValue {
+        value: fetched_value,
+        last_update: now(),
+    };
+
+    db.autocommit(
+        |dbtx, _pd| {
+            let key_bytes = db_key.consensus_encode_to_vec();
+            let value_bytes = db_value.consensus_encode_to_vec();
+            Box::pin(async move { dbtx.raw_insert_bytes(&key_bytes, &value_bytes).await })
+        },
+        None,
+    )
+    .await
+    .expect("DB operation cannot fail");
+
+    Ok(())
+}
+
+/// Service that keeps track of a fixed list of dynamic config fields. These are
+/// regularly re-fetched from the server and saved in the DB. Use
+/// `[DynConfigService::get_config_value]` to retrieve the latest value.
+#[derive(Debug)]
+pub struct DynConfigService {
+    db_key_prefix: u8,
+    tracked_values: BTreeSet<&'static str>,
+}
+
+impl DynConfigService {
+    // TODO: add example
+    pub fn builder() -> DynConfigServiceBuilder {
+        DynConfigServiceBuilder {
+            tracked_values: Vec::new(),
+            db: None,
+            db_key_prefix: None,
+        }
+    }
+
+    /// Returns the last fetched config value from the database. If the value
+    /// wasn't fetched yet or isn't decodable returns `None`.
+    pub async fn get_config_value<T: DynamicConfigValue>(
+        &self,
+        dbtx: &mut DatabaseTransaction<'_>,
+    ) -> Option<T> {
+        assert!(
+            self.tracked_values.contains(&T::FETCH_METHOD),
+            "You need to register the dynamic config value first before reading it"
+        );
+
+        let db_key = DynConfigKey {
+            db_key_prefix: self.db_key_prefix,
+            fetch_method: T::FETCH_METHOD.to_owned(),
+        };
+
+        let db_value = dbtx
+            .raw_get_bytes(&db_key.consensus_encode_to_vec())
+            .await
+            .expect("DB read can't fail")?;
+
+        let db_value = DynConfigValue::consensus_decode(
+            &mut db_value.as_slice(),
+            &ModuleDecoderRegistry::default(),
+        )
+        .expect("DB value should always be valid");
+
+        match serde_json::from_value(db_value.value) {
+            Ok(v) => Some(v),
+            Err(e) => {
+                error!(?e, "Failed to deserialize dynamic config value");
+                None
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone, Encodable, Decodable)]
+struct DynConfigKey {
+    db_key_prefix: u8,
+    fetch_method: String,
+}
+
+#[derive(Debug, Clone)]
+struct DynConfigValue {
```

Comment:
If these were in some `db.rs` sub-module, it would be easier to tell there are different from the rest, and could be qualified in the code with `db::` for name clashes.

I also wonder if not using word "dynamic" altogether would be a good idea to avoid all the confusion with other dyn-stuff. The fact that these config values are dynamic is not even all that important, IMO. Something from: "module*", "cached*", "extended", or something/nothing.

Reply:
I went with `ExtraConfigService`